packages/curve-ui-kit/src/shared/ui/DataTable.tsx

```diff
@@ -0,0 +1,158 @@
+import { useIntersectionObserver } from 'ui'
+import {
+  Cell,
+  Column,
+  ColumnDef,
+  flexRender,
+  getCoreRowModel,
+  getSortedRowModel,
+  Header,
+  Row,
+  SortingState,
+  useReactTable,
+} from '@tanstack/react-table'
+import { useSortFromQueryString } from '../../hooks/useSortFromQueryString'
+import { ArrowDownIcon } from '../icons/ArrowDownIcon'
+
+const { Sizing, Spacing, MinWidth } = SizesAndSpaces
+
+const getAlignment = <T extends unknown>({ columnDef }: Column<T>) =>
+  columnDef.meta?.type == 'numeric' ? 'right' : 'left'
+
+const DataCell = <T extends unknown>({ cell }: { cell: Cell<T, unknown> }) => {
+  const column = cell.column
+  return (
+    <Typography
+      variant={column.columnDef.meta?.variant ?? 'tableCellMBold'}
+      color="text.primary"
+      component="td"
+      sx={{
+        textAlign: getAlignment(column),
+        paddingInline: Spacing.sm,
+        paddingBlock: Spacing.md,
+        ...getExtraColumnPadding(column),
+      }}
+      data-testid={`data-table-cell-${column.id}`}
+    >
+      {flexRender(column.columnDef.cell, cell.getContext())}
+    </Typography>
+  )
+}
+
+const DataRow = <T extends unknown>({ row }: { row: Row<T> }) => {
+  const ref = useRef<HTMLTableRowElement>(null)
+  const entry = useIntersectionObserver(ref, { freezeOnceVisible: true })
+  return (
+    <TableRow
+      sx={{
+        marginBlock: 0,
+        height: Sizing['3xl'],
+        borderBottom: '1px solid',
+        borderColor: (t) => t.design.Layer[1].Outline,
+      }}
+      ref={ref}
+      data-testid={`data-table-row-${row.id}`}
+    >
+      {/* render cells when visible vertically, so content is lazy loaded */}
+      {entry?.isIntersecting && row.getVisibleCells().map((cell) => <DataCell key={cell.id} cell={cell} />)}
+    </TableRow>
+  )
+}
+
+/**
+ * In the figma design, the first and last columns seem to be aligned to the table title.
+ * However, the normal padding causes them to be misaligned.
+ */
+const getExtraColumnPadding = <T extends any>(column: Column<T>) => ({
+  ...(column.getIsFirstColumn() && { paddingInlineStart: Spacing.md }),
+  ...(column.getIsLastColumn() && { paddingInlineEnd: Spacing.md }),
+})
+
+const HeaderCell = <T extends unknown>({ header }: { header: Header<T, unknown> }) => {
+  const { column } = header
+  const sort = column.getIsSorted()
+  return (
+    <TableCell
+      sx={{
+        textAlign: getAlignment(column),
+        alignContent: 'end',
+        padding: Spacing.sm,
+        paddingBlockStart: 0,
+        ...getExtraColumnPadding(column),
+        ...(column.columnDef.enableSorting && {
+          cursor: 'pointer',
+        }),
+      }}
+      colSpan={header.colSpan}
+      width={header.getSize()}
+      onClick={column.getToggleSortingHandler()}
+      data-testid={`data-table-header-${column.id}`}
+    >
+      <Typography variant="tableHeaderS" color={`text.${sort ? 'primary' : 'secondary'}`}>
```

Comment:
Not part of this PR perhaps, but
```tsx
      <Typography
        variant="tableHeaderS"
        color={`text.${sort ? 'primary' : 'secondary'}`}
        sx={{
          ...(column.getCanSort() && {
            '&:hover': {
              color: 'text.primary',
            },
          }),
        }}
      >
```
would add some nice highlighting when you hover clickable table header cells. This isn't the right place, probably better suited in the default style configuration of header cells. It's a "nice to have".